File: 3-HARD/Python/3-suffix-array-construction/proposed-solution/suffix_array.py
from typing import List

class SuffixArray:
    def __init__(self, text: str):
        self.text = text
        self.n = len(text)
        self.suffix_array = []
        self.lcp_array = []

    def build_suffix_array(self) -> List[int]:
        self.suffix_array = sorted(range(self.n), key=lambda i: self.text[i:])
        return self.suffix_array

    def build_lcp_array(self) -> List[int]:
        if not self.suffix_array:
            self.build_suffix_array()

        self.lcp_array = [0] * self.n
        inv_suffix = [0] * self.n

        for i, suffix in enumerate(self.suffix_array):
            inv_suffix[suffix] = i

        k = 0
        for i in range(self.n):
            if inv_suffix[i] == self.n - 1:
                k = 0
                continue
            j = self.suffix_array[inv_suffix[i] + 1]
            while i + k < self.n and j + k < self.n and self.text[i + k] == self.text[j + k]:
                k += 1
            self.lcp_array[inv_suffix[i]] = k
            if k:
                k -= 1

        return self.lcp_array

File: 3-HARD/JavaScript/3-suffix-array-construction/proposed-solution/SuffixArray.js
// SuffixArray.js - Suffix Array and LCP Array construction in JavaScript

class SuffixArray {
  constructor(text) {
    this.text = text;
    this.n = text.length;
    this.suffixArray = [];
    this.lcpArray = [];
  }

  buildSuffixArray() {
    this.suffixArray = Array.from({ length: this.n }, (_, i) => i);
    this.suffixArray.sort((a, b) => {
      const substrA = this.text.substring(a);
      const substrB = this.text.substring(b);
      if (substrA < substrB) return -1;
      if (substrA > substrB) return 1;
      return 0;
    });
    return this.suffixArray;
  }

  buildLCPArray() {
    if (this.suffixArray.length === 0) {
      this.buildSuffixArray();
    }

    this.lcpArray = new Array(this.n).fill(0);
    const invSuffix = new Array(this.n);

    for (let i = 0; i < this.n; i++) {
      invSuffix[this.suffixArray[i]] = i;
    }

    let k = 0;
    for (let i = 0; i < this.n; i++) {
      if (invSuffix[i] === this.n - 1) {
        k = 0;
        continue;
      }
      const j = this.suffixArray[invSuffix[i] + 1];
      while (i + k < this.n && j + k < this.n && this.text[i + k] === this.text[j + k]) {
        k++;
      }
      this.lcpArray[invSuffix[i]] = k;
      if (k > 0) k--;
    }

    return this.lcpArray;
  }
}

export default SuffixArray;
